lib/elixir/lib/enum.ex

```diff
@@ -3492,6 +3494,279 @@ defmodule Enum do
+
+  ### Zipping Maps
+
+  It's important to remember that zipping inherently relies on order. If you zip two lists you get
+  the element at the index from each list in turn. If we zip two maps together it's tempting to
+  think that you will get the given key in the left map and the matching key in the right map, but
+  there is no such guarantee because map keys are not ordered! Consider the following:
+
+  ```elixir
+  left =  %{:a => 1, 1 => 3}
+  right = %{:a => 1, :b => :c}
+  Enum.zip(left, right)
+  # [{{1, 3}, {:a, 1}}, {{:a, 1}, {:b, :c}}]
+  ```
+  As you can see `:a` does not get paired with `:a`. If this is what you want, you should use
+  `Map.merge/3`
+
+  ### Examples
+
+      iex> Enum.zip_reduce([1, 2], [3, 4], 0, fn x, y, acc -> x + y + acc end)
+      10
+
+      iex> Enum.zip_reduce([1, 2], [3, 4], [], fn x, y, acc -> [x + y |acc] end)
+      [6, 4]
+  """
+  def zip_reduce(left, right, acc, reducer) when is_list(left) and is_list(right) do
+    non_stop_reducer = &{:cont, reducer.(&1, &2, &3)}
+
+    fast_enumerable_zip_reduce_while(left, right, {:cont, acc}, non_stop_reducer)
+    |> elem(1)
+  end
+
+  def zip_reduce(left, right, acc, reducer) do
+    non_stop_reducer = fn [l, r], acc -> {:cont, reducer.(l, r, acc)} end
+    enumerable_zip_reduce_while([left, right], {:cont, acc}, non_stop_reducer) |> elem(1)
+  end
+
+  @doc """
+  Reduces a over all of the given enums, halting as soon as either enumerable is empty.
+
+  An element is taken from each enum in turn and is passed to the reducer as a list.
+
+  ### Zipping Maps
+
+  It's important to remember that zipping inherently relies on order. If you zip two lists you get
+  the element at the index from each list in turn. If we zip two maps together it's tempting to
+  think that you will get the given key in the left map and the matching key in the right map, but
+  there is no such guarantee because map keys are not ordered! Consider the following:
+
+  ```elixir
+  left =  %{:a => 1, 1 => 3}
+  right = %{:a => 1, :b => :c}
+  Enum.zip(left, right)
+  # [{{1, 3}, {:a, 1}}, {{:a, 1}, {:b, :c}}]
+  ```
+  As you can see `:a` does not get paired with `:a`. If this is what you want, you should use
+  `Map.merge/3`
+
+  ### Examples
+
+      iex> enums = [[1, 1], [2, 2], [3, 3]]
+      ...>  Enum.zip_reduce(enums, [], fn elements, acc ->
+      ...>    [List.to_tuple(elements) | acc]
+      ...> end)
+      [{1, 2, 3}, {1, 2, 3}]
+
+      iex> enums = [[1, 2], %{a: 3, b: 4}, [5, 6]]
+      ...> Enum.zip_reduce(enums, [], fn elements, acc ->
+      ...>   [List.to_tuple(elements) | acc]
+      ...> end)
+      [{2, {:b, 4}, 6}, {1, {:a, 3}, 5}]
+  """
+  def zip_reduce(enums, acc, reducer) do
+    # turn the enums into Enumerable fns that take consume the thing off of it
+    non_stop_reducer = &{:cont, reducer.(&1, &2)}
+    enumerable_zip_reduce_while(enums, {:cont, acc}, non_stop_reducer) |> elem(1)
```

Comment:
Instead of implementing our own traversal, you should do like we do in `zip`, which is building a stream zip and then traversing. This should considerably simplify the implementation:

```elixir
   Stream.zip(enumerables).({:cont, []}, &{:cont, [&1 | &2]})
    |> elem(1)
    |> :lists.reverse()
```

Reply:
Right. Haha, well that's even quicker than what I did somehow, though I thought my traverse did less reversing!

I'll try that then.